cmd/snap-confine/snap-confine-invocation.c

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "snap-confine-invocation.h"
+
+#include <string.h>
+
+#include "../libsnap-confine-private/snap.h"
+#include "../libsnap-confine-private/utils.h"
+
+void sc_init_invocation(sc_invocation *inv, const struct sc_args *args, const char *snap_instance) {
+    /* Snap instance name is conveyed via untrusted environment. It may be
+     * unset (typically when experimenting with snap-confine by hand). It
+     * must also be a valid snap instance name. */
+    if (snap_instance == NULL) {
+        die("SNAP_INSTANCE_NAME is not set");
+    }
+    sc_instance_name_validate(snap_instance, NULL);
+
+    /* The security tag is conveyed via untrusted command line. It must be
+     * in agreement with snap instance name and must be a valid security
+     * tag. */
+    const char *security_tag = sc_args_security_tag(args);
+    if (!verify_security_tag(security_tag, snap_instance)) {
+        die("security tag %s not allowed", security_tag);
+    }
+
+    /* The base snap name is conveyed via untrusted, optional, command line
+     * argument. It may be omitted where it implies the "core" snap is the
+     * base. */
+    const char *base_snap_name = sc_args_base_snap(args) ?: "core";
```

Comment:
Yes this is a gcc. extension but it is convenient. I don't see why we should avoid using it.

Memory management is such that the values here are not freed since they are a bound the the lifecycle of `args`. I'm happy to explore changing this but this is done consistently across sc_invocation.

Reply:
I changed it though I hold my position that this is not an issue.